test/e2e/prometheus_test.go

```diff
@@ -3905,6 +3905,100 @@ func testPromEnforcedNamespaceLabel(t *testing.T) {
 	}
 }
 
+func testPrometheusCRDValidation(t *testing.T) {
+	t.Parallel()
+	name := "test"
+	replicas := int32(1)
+	commonFields := monitoringv1.CommonPrometheusFields{
+		Replicas: &replicas,
+		Version:  operator.DefaultPrometheusVersion,
+
+		ServiceAccountName: "prometheus",
+		Resources: v1.ResourceRequirements{
+			Requests: v1.ResourceList{
+				v1.ResourceMemory: resource.MustParse("400Mi"),
+			},
+		},
+	}
+
+	tests := []struct {
+		name           string
+		prometheusSpec monitoringv1.PrometheusSpec
+		expectedError  bool
+	}{
+		//
+		// RetentionSize Validation:
+		//
+		{
+			name: "zero-size-without-unit",
+			prometheusSpec: monitoringv1.PrometheusSpec{
+				CommonPrometheusFields: commonFields,
+				RetentionSize:          "0",
+			},
+		},
+		{
+			name: "legacy-unit",
+			prometheusSpec: monitoringv1.PrometheusSpec{
+				CommonPrometheusFields: commonFields,
+				RetentionSize:          "1.5GB",
+			},
+		},
+		{
+			name: "iec-unit",
+			prometheusSpec: monitoringv1.PrometheusSpec{
+				CommonPrometheusFields: commonFields,
+				RetentionSize:          "100MiB",
+			},
+		},
+		{
+			name: "legacy-missing-symbol",
+			prometheusSpec: monitoringv1.PrometheusSpec{
+				CommonPrometheusFields: commonFields,
+				RetentionSize:          "10M",
+			},
+			expectedError: true,
+		},
+		{
+			name: "legacy-missing-unit",
+			prometheusSpec: monitoringv1.PrometheusSpec{
+				CommonPrometheusFields: commonFields,
+				RetentionSize:          "1000",
+			},
+			expectedError: true,
+		},
+		{
+			name: "iec-missing-symbol",
+			prometheusSpec: monitoringv1.PrometheusSpec{
+				CommonPrometheusFields: commonFields,
+				RetentionSize:          "15Gi",
+			},
+			expectedError: true,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			testCtx := framework.NewTestCtx(t)
+			defer testCtx.Cleanup(t)
+			ns := framework.CreateNamespace(context.Background(), t, testCtx)
+			framework.SetupPrometheusRBAC(context.Background(), t, testCtx, ns)
+
+			prom := framework.MakeBasicPrometheusCRForValidation(ns, name, test.prometheusSpec)
+			prom.Namespace = ns
+
+			_, err := framework.CreatePrometheusAndWaitUntilReady(context.Background(), ns, prom)
+
+			if test.expectedError && err == nil {
+				t.Fatalf("expected error but got nil")
+			}
+			t.Logf("%v", err)
+		})
```

Comment:
Don't we miss the opposite check (e.g. non-nil error returned but not expected)?
We could also ensure that the returned error is about invalid input using  [`errors.IsInvalid`](https://pkg.go.dev/k8s.io/apimachinery@v0.23.5/pkg/api/errors#IsInvalid)).

```suggestion
			p := framework.MakeBasicPrometheusCRForValidation(ns, name, test.prometheusSpec)
			if tc.expectedError {
				_, err := framework.MonClientV1.Prometheuses(ns).Create(ctx, p, metav1.CreateOptions{})
				if err == nil {
					t.Fatal("expected error but got nil")
				}
				if !errors.IsInvalid(err) {
					t.Fatalf("expected Invalid error but got %v", err)
				}
				return
			}

			_, err := framework.CreatePrometheusAndWaitUntilReady(context.Background(), ns, prom)

			if err != nil {
				t.Fatalf("expected no error but got %v", err)
			}
		})
```